### PR TITLE
Fix block race

### PR DESCRIPTION
**Pull request by maackle**
### Summary

Nodes can easily get temporarily blocked and have their connections closed when first joining due to a race condition. This works around the race by ignoring an uninitialized node agent map.

~~Note that this PR is based on develop-0.2, because I was testing that version locally. Not sure if we want to merge to 0.2 or 0.3 or both.~~

### TODO:
- [ ] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: crates/holochain/tests/sharded_gossip/mod.rs [hunk 1]
```diff
@@ -201,6 +201,51 @@ async fn test_gossip_shutdown() {
     assert_eq!(record.unwrap().action_address(), &hash);
 }
 
+/// Test that when a new conductor joins, gossip picks up existing data without needing a publish.
+#[cfg(feature = "slow_tests")]
+#[tokio::test(flavor = "multi_thread")]
+async fn test_gossip_startup() {
```
Comment (maackle) [inline crates/holochain/tests/sharded_gossip/mod.rs:207]: This test passed even before the fix. It could be useful, but I think I'll just ignore it, since it does sleep.

File: crates/holochain/src/conductor/space.rs [hunk 2]
```diff
@@ -241,6 +241,12 @@ impl Spaces {
             }
         };
 
+        // If node_agents_in_spaces is not yet initialized, we can't know anything about
+        // which cells are blocked, so avoid the race condition by returning false
+        if cell_ids.is_empty() {
+            return Ok(false);
+        }
```
Comment (neonphog) [inline crates/holochain/src/conductor/space.rs:248]: Okay merging this to fix the race condition. But, security-wise, is there any other case that might trigger this early exit that we actually *would* want the connection to be blocked? Such as all cells being stopped or uninstalled, or something?

**Review by neonphog — APPROVED** [review]

File: crates/holochain/tests/sharded_gossip/mod.rs [hunk 3]
```diff
@@ -201,6 +201,53 @@ async fn test_gossip_shutdown() {
...
+
+    // Startup and do peer discovery
+    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+    let (conductor1, cell1, zome1) = mk_conductor().await;
+
+    // Wait a bit so that conductor 0 doesn't publish in the next step.
+    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+    SweetConductor::exchange_peer_info([&conductor0, &conductor1]).await;
+
+    consistency_60s([&cell0, &cell1]).await;
```
Comment (jost-s) [inline crates/holochain/tests/sharded_gossip/mod.rs:246]: this seems very long, shouldn't gossip start directly after new peers were discovered? is that why the test passed before the fix?

**Review by jost-s — APPROVED** [review]

File: crates/holochain/src/conductor/space.rs (hunk 2 above)
Comment (maackle) [reply crates/holochain/src/conductor/space.rs:248]: Well it seems the only way there can be no agents is if the p2p db has no agents that match the node id. This is probably happening because alice discovered bob before bob discovered alice. Bob doesn't have the info necessary to determine whether he has blocked alice, he won't know that until he discovers her.

@thedavidmeister @neonphog can you help think through this? What are the situations where bob can not know alice's agent info? Certainly when alice first joins, and in that case we should probably just let them talk. But if bob blocked alice previously, would he ever forget her agent info and run into this case?

Another option is to make this check return an Err, which will trigger a `MetaNetEvtAuth::UnauthorizedIgnore`. That might be more correct, but it would probably gunk up the works when a new agent joins if we don't fix the race condition, because then a new agent is likely to be left hanging by whoever they talk to, depending on who wins the race.

As this PR stands, it fixed the slow gossip issue I was seeing. But yes we need to make sure this isn't some kind of security loophole.

Comment (neonphog) [reply crates/holochain/src/conductor/space.rs:248]: @maackle - a bit of info you don't have i think is the "preflight" connection message which automatically exchanges all local agent infos for the space on connection establishment before the ban check is made.

There was a test that showed that this worked properly, but of course there could be corner cases not covered.

File: crates/holochain/tests/sharded_gossip/mod.rs (hunk 3 above)
Comment (maackle) [reply crates/holochain/tests/sharded_gossip/mod.rs:246]: gossip actually succeeds in less than 3 seconds so this timeout is not helping it pass

File: crates/holochain/src/conductor/space.rs (hunk 2 above)
Comment (maackle) [reply crates/holochain/src/conductor/space.rs:248]: I'll add a warning log about this when the agent list is empty, which will let me verify that the preflight is indeed not working, and will also help us follow up with it later.

Comment (maackle) [reply crates/holochain/src/conductor/space.rs]: Yep, here's what I see:

If I start up conductor 1 first, commit some entries, and then spin up conductor 2 on the same network, as soon as conductor 2 joins, conductor 1's log contains 18 instances in quick succession of the warning I just put in about an empty agent list. After doing some more stuff, including stopping and starting nodes, I haven't seen this warning come up again. However I do see some scattered instances of "No agents to gossip with on the node". But gossip performance seems to be good, and no more closed connections.

File: crates/holochain/src/conductor/space.rs [hunk 4]
```diff
@@ -241,6 +241,13 @@ impl Spaces {
             }
         };
 
+        // If node_agents_in_spaces is not yet initialized, we can't know anything about
+        // which cells are blocked, so avoid the race condition by returning false
+        if cell_ids.is_empty() {
+            tracing::warn!(?target_id, "is_blocked(): No agents were found for the target. This indicates that preflight agent exchange may not be working properly.");
```
Comment (thedavidmeister) [inline crates/holochain/src/conductor/space.rs:247]: exactly... shouldn't we fix the preflight if it's broken and causing problems?

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

**Review by thedavidmeister — APPROVED** [review]
approved on the assumption that we do actually fix the real issue

File: crates/holochain/src/conductor/space.rs (hunk 4 above)
Comment (ThetaSinner) [reply crates/holochain/src/conductor/space.rs:247]: Do we know that Holochain recovers from this situation? Or is is possible that it stays in this state?
